README.md
# Pyre Bot

[![made-with-python](https://img.shields.io/badge/Made%20with-Python-1f425f.svg)](https://www.python.org/)[![DeepSource](https://static.deepsource.io/deepsource-badge-light-mini.svg)](https://deepsource.io/gh/InfernalPlacebo/pyre-bot/?ref=repository-badge)![Discord](https://img.shields.io/discord/660914305515913227?label=Discord)[![GPLv3 license](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://github.com/InfernalPlacebo/ig-bot/blob/master/LICENSE)[![Documentation Status](https://readthedocs.org/projects/pyre-bot/badge/?version=latest)](https://pyre-bot.readthedocs.io/en/latest/?badge=latest)

Please find complete documentation on [ReadTheDocs](https://docs.pyre-bot.com)

The Pyre Discord Bot has one main goal: Create an easy-to-use way to manage a maintain dedicated game servers from within Discord. From sending commands to the game console to allowing players to initiate a restart vote, Pyre was created to make it as easy as possible.

Feel free to join our [Discord](http://discord.pyre-bot.com) if you want to discuss the bot or hang out!

## Getting Started

These instructions will help make sure your system is ready to run the bot and help you get started using it for your needs.

### Prerequisites

The bot is built upon Python 3.8 but has been tested working with Python 3.7 without any issues. The dependencies for the usage are updated regularly in the requirements.txt file and these can be installed via:

```
pip install -r requirements.txt
```

As of writing the current requirements are:

```
boto3==1.12.37
python_a2s==1.1.4
discord.py==1.3.3
psutil==5.6.7
requests==2.22.0
python_valve==0.2.1
discord==1.0.1
valve==0.0.0
```

If you get an aiohttp error while attempting to install discord.py, run the following command:

```
pip install --no-use-pep517 discord.py
```

### Installing

Clone the repo to wherever you want the bot to reside. You can run the bot by calling **bot.py**. You will also need [SteamCMD](https://developer.valvesoftware.com/wiki/SteamCMD) installed on your machine to run the updates for the game servers. When the bot first runs it will open a setup window where it prompts you for configuration settings. The window will also allow you to install the BotCommands plugin, this is required if you want access to functions that utilize in-game commands from Discord.

#### API Tokens

##### Discord
You need to get an API token from the [Discord Developer Portal](https://discordapp.com/developers/docs/intro). The token is added to the config.ini file in the config folder.

##### AWS
Pyre Bot uses Amazon DynamoDB to store player stats from the servers and link Discord IDs to SteamIDs. You will need to setup an AWS account an add some tables to DynamoDB for this to work. Create the following tables:

* **Discord_Stats** - Primary key: DiscordID (String)
* **Players** - Primary key: DiscordID (String)
* **Stats** - Primary key: SteamID64 (String)

You also need to setup AWS credentials on your computer. You can use [AWS CLI](https://aws.amazon.com/cli/) to do this.

#### Setting variables
Some variables need to be set before using the bot to make sure it is looking in the correct place for files and information. These variables are setup when the bot first runs when it opens the setup window. In the **config/config.ini** file you will be able to config the following:

```
[API]
discord_token = token

[General]
role = privileged-server-role
admin-channel = admin-channel-id
commands-channel = commands-channel-id
linked-id = linked-role-id

[RoR2]
server_address = your-server-address
server_port = your-server-port
steamcmd = path-to-steamcmd
ror2ds = path-to-ror2ds
bepinex = path-to-bepinex
channel = enter-channel-id
auto-start-chat = true
auto-server-restart = true
hidden_mods = hidden-mods-here
```

* **discord_token**: The API key you retrieved in the earlier step
* **role**: The Discord role you want using protected commands
* **admin-channel**: The channel ID that is used to issue admin commands
* **linked-id**: The id of the "Linked" Discord role
* **commands-channel**: The channel ID that anyone can use to issue commands
* **server_address**: the IP/domain of your server
* **server_port**: The port configured for queries
* **steamcmd**: The path to the steamcmd folder
* **ror2ds**: Used for ror2.py, path to the Risk of Rain 2 Dedicated Server folder
* **BepInEx**: Path to the BepInEx folder
* **channel**: The Discord channel ID to output live chat
* **auto-start-chat**: Set to false to prevent the bot from outputting chat to a Discord channel when it launches
* **auto-server-restart**: Set to false to prevent the bot from restarting the server every 60 minutes in which no players join the lobby
* **hidden_mods**: Add mods that you don't want to be listed by the mods command, ships with a default list

*Never upload online or share your config file to anyone you do not trust. These API keys are private and can result in your access from the services being removed if they get out.*

#### Risk of Rain 2 requirements

The bot requires that you are using the following mods in your RoR2 server. Not having all of these mods installed and updated to their latest builds can result in certain features not working properly.

**Required Mods**
* [BepInEx](https://thunderstore.io/package/bbepis/BepInExPack/)
  * Change **redirectOutputLog** to **true** in your doorstop_config.ini to prevent double messages being sent to BepInEx terminal.
* [BotCommands](https://github.com/SuperRayss/BotCommands)
  * Used to send commands to the server from Discord
* [R2DSE](https://thunderstore.io/package/Harb/R2DSEssentials/)
  * Outputs steam player names and IDs
* [DebugToolkit (Custom Fork by Rayss)](https://github.com/SuperRayss/DebugToolkit)
  * Outputs run time and stages cleared
  * Enables additional commands to be sent to the server (i.e. give_item, give_equip)

## Running and using the bot

The bot can be used by running **bot.py** in the main directory. The bot will output a message stating that it is connected to Discord and ready to listen when it starts. Once you get the confirmation message you are able to start issuing commands to your bot using one of the command prefixes ('r!', 'ig!', '>').

## Bot Commands

### Available to everyone
* restart : Initiates a vote to restart the server
  * Requires a majority vote (at least 75% of in-game player count)
* votekick {player}: Initiates a 30-second vote to kick {player} from the Risk of Rain 2 server
  * Requires a majority vote (at least 75% of in-game player count)
* endrun : Initiates a 30-second vote to end the current run in the Risk of Rain 2 server
  * Requires a majority vote (at least 75% of in-game player count)
* info : Lists the game server status via Steamworks API
  * Listed info includes server name, current stage, player count, player names, and ping
* link {SteamID} : Allows players to see their play stats for the Risk of Rain 2 server
  * Assigns the "Linked" role (NOTE: requires further config to not be hardcoded)
* stats : Outputs your player stats to chat
  * Currently supported stats are Time Played, Stages Cleared, and Runs Completed
* mods : Outputs a list of mods to chat
* config : Outputs the current server config to chat
  * Coming soon!

### Admin only
* start : Starts the Risk of Rain 2 server
* stop : Stops the Risk of Rain 2 server
* update : Updates the Risk of Rain 2 server
* delete {number} : Deletes the given amount of messages in the channel
* start_chat : Reads the Risk of Rain 2 server logs and outputs live chat from the game to a specified Discord chat channel
* stop_chat : Stops reading out Risk of Rain 2 server chat
* say {message} : Sends an in-game message from the perspective of the server
* cmd {command with args} : Passes on a command to execute by the server
  * Very experimental as of now, only use when you are sure of the results, as passing commands in certain contexts can cause unhandled exceptions with BotCommands
* giveitem {player} {item} {quantity (default - 1)} : Gives a player a specified quantity of an item
* giveequip {player} {equip} : Gives a player a specified equipment

## Modifying the commands

By default the bot is configured to manage a Risk of Rain 2 server. This can be changed by calling the executables for other servers and pointing the checks at their respective logs.

## Built With

* [Discord.py](https://github.com/Rapptz/discord.py) - Discord API wrapper for Python
* [Python-A2S](https://github.com/Yepoleb/python-a2s) - Steamworks API wrapper for Python

## Contributing

Please read [CONTRIBUTING.md](https://github.com/InfernalPlacebo/pyre-bot/blob/master/CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests to us.

To add functionality for other games or servers, add a new file in the **cogs** folder.

## Authors

* **Wade Fox** - *Creator* - [GitHub](https://github.com/InfernalPlacebo), [Discord](http://discord.pyre-bot.com)
* **Rayss** - *Contributor and BotCommands creator* - [GitHub](https://github.com/SuperRayss), [Discord](http://discord.pyre-bot.com)

See also the list of [contributors](https://github.com/InfernalPlacebo/pyre-bot/graphs/contributors) who participated in this project.

## License

This project is licensed under the GPL-3.0 License - see the [LICENSE](LICENSE) file for details

## Changelog

### 0.8.0
* Moved all stat tracking to Amazon DynamoDB for a better experience
* Custom version of the Pygtail module is now packaged with the bot
* Fixed an exception that occurred when the bot reconnected to Discord servers.
* Complete rewrite of setup.py
* Updated all requirements to newest versions
* Various fixes, improvements, and optimizations that makes everyone's lives better

### 0.7.0
* Added player stat tracking
* Improved integration with DebugToolkit
* Gets run time and stage number from the server log output directly
* Other things I've likely just forgotten about ... it's been a while

### 0.6.0
* Added configuration options for admin and command channels
  * This allows running multiple bot instances to manage more than one server
  * New global check makes sure the bot only listens to commands in these channels
* Added dictionaries for items, equipment, and stages
* Commands check dictionaries for values before issuing to the server
* Added basic logging
 * Log can be found in the **bot.log** file in the bot's directory.
* New *delete* command allows simpler delete or testing messages
* Improved error handling
* Steam query is now used to check if the server is running

### 0.5.1
* Added *giveequip* command
* *status* command now shows current stage name

### 0.5.0
* Added *say*, *endrun*, *giveitem*, and *votekick* commands
  * Requires [BotCommands](https://github.com/SuperRayss/BotCommands)
* Added a setup script to create config.ini on first launch
* Added option to output game chat to Discord channel
  * Enabled by default, change in config.ini
* Added automatic server restarts
  * Enabled by default, change in config.ini
* Using new Steamworks API


docs/CHANGELOG.md
## Changelog

### 0.8.0
* Moved all stat tracking to Amazon DynamoDB for a better experience
* Custom version of the Pygtail module is now packaged with the bot
* Fixed an exception that occurred when the bot reconnected to Discord servers.
* Complete rewrite of setup.py
* Updated all requirements to newest versions
* Various fixes, improvements, and optimizations that makes everyone's lives better

### 0.6.0
* Added configuration options for admin and command channels
  * This allows running multiple bot instances to manage more than one server
  * New global check makes sure the bot only listens to commands in these channels
* Added dictionaries for items, equipment, and stages
* Commands check dictionaries for values before issuing to the server
* Added basic logging
 * Log can be found in the **bot.log** file in the bot's directory.
* New *delete* command allows simpler delete or testing messages
* Improved error handling
* Steam query is now used to check if the server is running

### 0.5.1
* **Name changed to Pyre Bot**
* Added *giveequip* command
* *status* command now shows current stage name

### 0.5.0
* Added *say*, *endrun*, *giveitem*, and *votekick* commands
    * Requires [BotCommands](https://github.com/SuperRayss/BotCommands)
* Added a setup script to create config.ini on first launch
* Added option to output game chat to Discord channel
    * Enabled by default, change in config.ini
* Added automatic server restarts
    * Enabled by default, change in config.ini
* Using new Steamworks API

### 0.4.1
* Able to view server mods with the >mods commands
* ror2.py outputs load AND unload to terminal
* Misc code cleanup

### 0.4.0
* We welcome **Rayss** as a contributor to the project!
* Added config.ini
* New function to allow commands to output to a specified Discord channel
    * Requires the role specified in config.ini to use these commands
    * Discord channel is specified in config.ini
* ror2.py now outputs to terminal when it is loaded
* Minor fix to update command, removes log file *before* starting the update now
* Commands no longer require proper capitalization


### 0.3.2
* Thanks to Rayss for testing the bot and providing valuable feedback!
* Added a new variable for the Risk of Rain server location
* Removed link command until I can figure out how to fix it
* Fixed the status command to show the guild name vs being hard coded
* Load, unload, and reload commands can now only be used by the bot owner
* Changed restart vote to require 75% of the server player count (via Steamworks)

### 0.3.1
* Removed League of Legends support and dependencies... for now.

### 0.3.0
* Added the restart feature
    * Allows Discord members to initiate a restart vote for the RoR2 server.
* Changed the protected features check to a configurable variable
* Lowered wait time for start and stop commands to 15 seconds

### 0.2.0
* Added Cogs
    * Added cogs to the code to provide expandable in future releases and allow cleaner reading and writing of code.
    * Cogs allow development and test of features and changes without having to completely restart bot.


cogs/ror2.py
#!/usr/bin/env python3

"""Pyre Bot Risk of Rain 2 user functions."""

import ast
import asyncio
import logging
import os
from configparser import ConfigParser
from pathlib import Path

import a2s
import discord
import psutil
from discord.ext import commands

config_object = ConfigParser()
config_file = Path.cwd().joinpath('config', 'config.ini')
config_object.read(config_file)
ror2 = config_object["RoR2"]
general = config_object["General"]

# Config variables
server_address = config_object.get(
    'RoR2', 'server_address'), config_object.getint('RoR2', 'server_port')
steamcmd = Path(ror2["steamcmd"])
ror2ds = Path(ror2["ror2ds"])
BepInEx = Path(ror2["BepInEx"])
role = general["role"]
c_autostart = ror2['auto-start-chat']
s_restart = ror2['auto-server-restart']
hidden_mods = ast.literal_eval(config_object.get('RoR2', 'hidden_mods'))
botcmd = Path.joinpath(BepInEx, 'plugins', 'BotCommands')

# Global variables (yes, I know, not ideal but I'll fix them later)
yes, no = 0, 0

logfile = (BepInEx / "LogOutput.log")

# These get assigned / updated every time server() is called
# Only using string type as a placeholder to avoid exceptions if the server is not online when the bot initializes
server_info = ''
server_players = ''

# Dictionaries used for functions
stages = {
    'title': 'Title',
    'lobby': 'Game Lobby',
    'blackbeach': 'Distant Roost',
    'blackbeach2': 'Distant Roost',
    'golemplains': 'Titanic Plains',
    'golemplains2': 'Titanic Plains',
    'foggyswamp': 'Wetland Aspect',
    'goolake': 'Abandoned Aqueduct',
    'frozenwall': 'Rallypoint Delta',
    'wispgraveyard': 'Scorched Acres',
    'dampcave': 'Abyssal Depths',
    'shipgraveyard': "Siren's Call",
    'arena': 'Hidden Realm: Void Fields',
    'bazaar': 'Hidden Realm: Bazaar Between Time',
    'goldshores': 'Hidden Realm: Glided Coast',
    'mysteryspace': 'Hidden Realm: A Moment, Fractured',
    'limbo': 'Hidden Realm: A Moment, Whole',
    'artifactworld': 'Hidden Realm: Artifact World',
    'skymeadow': 'Sky Meadow'
}


async def server():
    """
    Checks if the server is running or not.
    Returns:
        Boolean: Used by functions calling this to check if running
    """
    global server_info
    global server_players
    try:
        server_info = a2s.info(server_address, 1.0)
        server_players = a2s.players(server_address)
        return True
    except:
        #        print("Server error:", sys.exc_info()[0], sys.exc_info()[1]) #  Used for debugging
        return False


async def server_stop():
    """
    Stops the server.
    Returns:
        Boolean: Indicates whether server stopped or not
    """
    for proc in psutil.process_iter():
        exe = Path.cwd().joinpath(ror2ds, 'Risk of Rain 2.exe')
        try:
            processExe = proc.exe()
            if str(exe) == processExe:
                proc.kill()
                logging.info('Server stopped')
                return True
        except:
            pass
    return False


async def find_dll():
    """
    Checks to see if the BotCommands plugin is installed on server.
    Returns:
        Boolean: If true it is, otherwise it is not
    """
    plugin_dir = (BepInEx / 'plugins')
    files = [file.name for file in plugin_dir.glob('**/*') if file.is_file()]
    if 'BotCommands.dll' in files:
        return True
    logging.warning('Unable to find BotCommands.dll!')
    return False


class RoR2(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    # Counts reactions of commands with votes
    @commands.Cog.listener()
    async def on_raw_reaction_add(self, payload):
        global yes, no
        if payload.emoji.name == "✅":
            yes = yes + 1
        elif payload.emoji.name == "❌":
            no = no + 1
        else:
            pass

    # Restart the server with votes
    @commands.command(
        name='restart',
        help='Initializes a vote to restart the RoR2 server',
        usage='time'
    )
    async def restart(self, ctx, time=15):
        if await server():
            logging.info(f'{ctx.message.author.name} used {ctx.command.name}')
            global yes, no
            yes, no = 0, 0
            author = ctx.author
            message = await ctx.send('A restart vote has been initiated by '
                                     + f'{author.mention}. Please react to this message'
                                     + ' with your vote!')
            for emoji in ('✅', '❌'):
                await message.add_reaction(emoji)
            await asyncio.sleep(time)
            # Counts vote, if tie does nothing
            if yes == no:
                logging.info('There were not enoough votes to restart the server')
                await ctx.send('It was a tie! There must be a majority to restart the '
                               + 'server!')
            # If 75% of player count wants to restart it will
            elif (yes - 1) >= (server_info.player_count * 0.75):
                started = 1
                stopped = await server_stop()
                if stopped is True:
                    await ctx.send('Risk of Rain 2 server shut down...')
                elif stopped is False:
                    await ctx.send('Unable to stop server!')
                await asyncio.sleep(5)

                # Path of log file, removes before starting
                if os.path.exists(BepInEx / "LogOutput.log"):
                    try:
                        os.remove(BepInEx / "LogOutput.log")
                    except Exception:
                        print('Unable to remove log file')

                # Starts the server
                os.startfile(ror2ds / "Risk of Rain 2.exe")
                await ctx.send('Starting Risk of Rain 2 Server, please wait...')
                await asyncio.sleep(15)

                # After 15 seconds checks logs to see if server started
                while started == 1:
                    with open(BepInEx / "LogOutput.log") as f:
                        for line in f:
                            if "Loaded scene lobby" in line:
                                await ctx.send('Server started successfully...')
                                started = 2
                                break
                # All other options
            else:
                logging.info('There were not enoough votes to restart the server')
                await ctx.send('Restart vote failed!')
        else:
            await ctx.send('Server is not running, unable to restart...')

    # Kick a player with a majority vote
    # TODO: Add the ability to call this command with in-game chat by adding a
    # conditional to the chat command, so players can do it while in-game too.
    # Would have to add functionality for votes to count with in-game chat
    # though. (or not, if I want to leave that to the discord).
    @commands.command(
        name='votekick',
        help='Begins a vote to kick a player from the game',
        usage='playername'
    )
    async def votekick(self, ctx, *, kick_player):
        if await server() and await find_dll() is True:
            global yes, no
            yes, no = 0, 0
            author = ctx.author
            time = 30
            containskickplayer = 0
            for player in server_players:
                if kick_player.upper() in player.name.upper():
                    containskickplayer = 1
                    kick_player = player.name
                    break
            if containskickplayer == 1:
                logging.info(
                    f'{ctx.message.author.name} started a vote to kick {kick_player}')
                message = await ctx.send('A vote to kick ' + kick_player
                                         + f' has been initiated by {author.mention}. '
                                         + 'Please react to this message with your '
                                         + 'vote!')
                for emoji in ('✅', '❌'):
                    await message.add_reaction(emoji)
                await asyncio.sleep(time)
                # Counts vote, if tie does nothing
                if yes == no:
                    await ctx.send(
                        'It was a tie! There must be a majority to kick '
                        + kick_player
                    )
                # If 75% of player count wants to kick it will
                elif (yes - 1) >= (server_info.player_count * 0.75):
                    logging.info(f'{kick_player} was kicked from the game.')
                    append = open(botcmd / "botcmd.txt", 'a')
                    append.write('ban "' + kick_player + '"\n')
                    append.close()
                    await ctx.send('Kicked player ' + kick_player)
                # If vote fails
                else:
                    logging.info('Not enough votes to pass')
                    await ctx.send('Vote failed. There must be a majority to kick '
                                   + kick_player
                                   )
            else:
                await ctx.send(kick_player + ' is not playing on the server')
        elif await server() is False:
            await ctx.send('Server is not running...')
        elif await find_dll() is False:
            await ctx.send('BotCommands plugin is not loaded on the server!')

    @votekick.error
    async def votekick_handler(self, ctx, error):
        if isinstance(error, commands.MissingRequiredArgument):
            if error.param.name == 'kick_player':
                await ctx.send('Please insert a partial or complete player name')

    # Ends the run with a majority vote
    # TODO: Add the ability to call this command with in-game chat by adding a
    # conditional to the chat command, so players can do it while in-game too.
    # Would have to add functionality for votes to count with in-game chat
    # though. (or not, if I want to leave that to the discord).
    @commands.command(
        name='endrun',
        help='Begins a vote to end the current run',
    )
    async def endrun(self, ctx):
        if await server() and await find_dll() is True:
            logging.info(f'{ctx.message.author.name} started an end run vote')
            global server_info
            if server_info.map_name in ('lobby', 'title'):
                await ctx.send('No run in progress.')
            else:
                global yes, no
                yes, no = 0, 0
                author = ctx.author
                time = 30
                message = await ctx.send('A vote to end the run has been initiated by '
                                         + f'{author.mention}. Please react to this message'
                                         + ' with your vote!')
                for emoji in ('✅', '❌'):
                    await message.add_reaction(emoji)
                await asyncio.sleep(time)
                # If 75% of player count wants to end the run it will
                if (yes - 1) >= (server_info.player_count * 0.75):
                    logging.info('Vote passed to end the current run')
                    append = open(botcmd / "botcmd.txt", 'a')
                    append.write('run_end' + '\n')
                    append.close()
                    await ctx.send('Run ended, all players have been returned to the lobby')
                # If vote fails
                else:
                    logging.info('End run vote failed')
                    await ctx.send('Vote failed. There must be a majority to end the run')
        elif await server() is False:
            await ctx.send('Server is not running...')
        elif await find_dll() is False:
            await ctx.send('BotCommands plugin is not loaded on the server!')

    # Displays the status of the server
    @commands.command(
        name='info',
        help='Displays Risk of Rain 2 server information'
    )
    async def status(self, ctx):
        logging.info(f'{ctx.message.author.name} used {ctx.command.name}')
        if await server():
            stage = '???'
            # Create embed
            embed = discord.Embed(
                title='Server Information',
                colour=discord.Colour.blue()
            )

            # Creates the string of player names used in the embed
            player_names = []
            for player in server_players:
                player_names.append(player.name)
            player_names = ("\n".join(map(str, player_names)))

            # Convert Steam map name to game name
            for key, value in stages.items():
                if key in server_info.map_name:
                    stage = value
                    break

            # Embed information
            embed.set_footer(
                text=f'Requested by {ctx.message.author.name}',
                icon_url=self.bot.user.avatar_url
            )
            embed.set_thumbnail(url=self.bot.user.avatar_url)
            embed.set_author(name=self.bot.guilds[0])
            embed.add_field(name='Server Name',
                            value=f'{server_info.server_name}', inline=False)
            embed.add_field(name='Current Stage', value=f'{stage}', inline=False)
            embed.add_field(
                name='Player Count',
                value=f'{server_info.player_count}/{server_info.max_players}', inline=False)
            if server_info.player_count == 0:
                pass
            else:
                embed.add_field(
                    name='Players', value=player_names, inline=False)
            embed.add_field(name='Server Ping',
                            value=int(server_info.ping * 1000), inline=False)

            # Send embed
            await ctx.send(embed=embed)
        else:
            await ctx.send('Server is currently offline.')

    # Send modlist to chat
    @commands.command(
        name='mods',
        help='Lists all the mods currently running on the server'
    )
    async def mods(self, ctx):
        logging.info(f'{ctx.message.author.name} used {ctx.command.name}')
        mods = []
        with open(BepInEx / "LogOutput.log") as f:
            for line in f:
                if "[Info   :   BepInEx] Loading" in line:
                    line = line[30:]
                    head, sep, tail = line.partition(' ')
                    if head in hidden_mods:
                        pass
                    else:
                        mods.append(head)
        mods = ("\n".join(map(str, mods)))
        mod_embed = discord.Embed(colour=discord.Colour.blue())
        mod_embed.set_footer(
            text=f'Requested by {ctx.message.author.name}',
            icon_url=self.bot.user.avatar_url
        )
        mod_embed.add_field(name='Mods', value=mods, inline=False)
        await ctx.send(embed=mod_embed)


def setup(bot):
    """Loads the cog into bot.py."""
    bot.add_cog(RoR2(bot))
    print('Loaded cog: ror2.py')


def teardown(bot):
    """Prints to terminal when cog is unloaded."""
    print('Unloaded cog: ror2.py')
